### PR TITLE
Refactor transfer component

### PR DESCRIPTION
**Pull request by cehali**
# Summary
Transfer component refactoring

### CODE REVIEW

**Comment by cehali** [comment]
image
image
image
image

**Comment by cehali** [comment]
image
image
image
image

File: universal-login-commons/lib/core/utils/isValidAmount.ts [hunk 1]
```diff
@@ -1,9 +1,11 @@
 import {utils} from 'ethers';
 
-export const isValidAmount = (amount: string, balance: string): boolean => {
-  if (amount.match(/(^[0-9]+(\.?[0-9])*$)/)) {
-    const amountAsBigNumber = utils.bigNumberify(utils.parseEther(amount));
-    return amountAsBigNumber.gt(0) && amountAsBigNumber.lt(utils.parseEther(balance));
+export const isValidAmount = (balance: string | null, amount?: string): boolean => {
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/isValidAmount.ts]: In general it is a bad practice to pass complex type like `string | false`;
It is better to return nullable:
```ts
type Nullable<T> = T | null; 

const resolveName = async (...): Nullable<string> => {
...
}

```

File: universal-login-commons/lib/core/utils/isValidAmount.ts [hunk 2]
```diff
@@ -1,9 +1,11 @@
...
+  if (amount && balance) {
+    if (amount.match(/(^[0-9]+(\.?[0-9])*$)/)) {
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/isValidAmount.ts]: extract const

File: universal-login-commons/lib/core/utils/isValidRecipient.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+import {isProperAddress} from './hexStrings';
+import {isValidEnsName} from './ens';
+
+export const isValidRecipient = (recipient?: string): boolean => {
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/isValidRecipient.ts]: :+1

File: universal-login-commons/test/core/utils/isValidAmount.ts [hunk 4]
```diff
@@ -2,20 +2,22 @@ import {expect} from 'chai';
...
-  itValidatesAmount(true, '10', '20');
-  itValidatesAmount(true, '2', '2.01');
-  itValidatesAmount(false, 'ABC', '1');
-  itValidatesAmount(false, '0', '1');
-  itValidatesAmount(false, '1a00.1', '1');
-  itValidatesAmount(false, '14$.9', '1');
-  itValidatesAmount(false, '100.', '1');
-  itValidatesAmount(false, '100', '1');
-  itValidatesAmount(false, '100', '99');
+  itValidatesAmount(true, '2', '1.0');
```
Comment (marekkirejczyk) [inline universal-login-commons/test/core/utils/isValidAmount.ts]: Better order of argumentns could be:
`itValidatesAmount('1.0', 2, true);`
#minor

File: universal-login-react/src/core/models/ReactDashboardContentType.ts [hunk 5]
```diff
@@ -1 +1 @@
-export type DashboardContentType = 'funds' | 'topup' | 'transferAmount' | 'transferRecipient' | 'waitingForTransfer' | 'devices' | 'backup' | 'none';
+export type DashboardContentType = 'funds' | 'topup' | 'transferAmount' | 'waitingForTransfer' | 'devices' | 'backup' | 'none';
```
Comment (marekkirejczyk) [inline universal-login-react/src/core/models/ReactDashboardContentType.ts]: Should we rename `transferAmount` to simply `transfer`?

Comment (cehali) [reply universal-login-react/src/core/models/ReactDashboardContentType.ts]: It turns out `DashboardContentType` is not used anywhere. Deleted.

File: universal-login-commons/lib/core/utils/isValidAmount.ts (hunk 1 above)
Comment (cehali) [reply universal-login-commons/lib/core/utils/isValidAmount.ts]: Done

File: universal-login-commons/lib/core/utils/isValidAmount.ts (hunk 2 above)
Comment (cehali) [reply universal-login-commons/lib/core/utils/isValidAmount.ts]: Done

File: universal-login-commons/test/core/utils/isValidAmount.ts (hunk 4 above)
Comment (cehali) [reply universal-login-commons/test/core/utils/isValidAmount.ts]: Done

File: universal-login-commons/lib/core/utils/isValidAmount.ts [hunk 6]
```diff
@@ -1,9 +1,14 @@
 import {utils} from 'ethers';
+import {Nullable} from '../types/common';
 
-export const isValidAmount = (amount: string, balance: string): boolean => {
-  if (amount.match(/(^[0-9]+(\.?[0-9])*$)/)) {
-    const amountAsBigNumber = utils.bigNumberify(utils.parseEther(amount));
-    return amountAsBigNumber.gt(0) && amountAsBigNumber.lt(utils.parseEther(balance));
+const isNumber = /(^[0-9]+(\.?[0-9])*$)/;
```
Comment (marekkirejczyk) [inline universal-login-commons/lib/core/utils/isValidAmount.ts]:
```suggestion
const isNumber = /^[0-9]+(\.[0-9]+)?$/;
```

File: universal-login-wallet/test/integration/ui/Transfer.tsx [hunk 7]
```diff
@@ -23,24 +23,46 @@ describe('UI: Transfer', () => {
...
     ({mockTokenContract} = await createFixtureLoader(provider as providers.Web3Provider)(deployMockToken));
     ({appWrapper, appPage, services} = await setupUI(relayer, mockTokenContract.address));
-  });
-
-  it('Creates wallet and transfers ethers', async () => {
     const walletAddress = services.walletService.state.kind === 'Deployed'
       ? services.walletService.getDeployedWallet().contractAddress
       : '0x0';
     await mockTokenContract.transfer(walletAddress, utils.parseEther('2.0'));
     appPage.dashboard().clickTransferButton();
```
Comment (marekkirejczyk) [inline universal-login-wallet/test/integration/ui/Transfer.tsx]: const = transferPage = appPage.dashboard().gotoTransferButton();

File: universal-login-commons/lib/core/utils/isValidRecipient.ts [hunk 8]
```diff
@@ -0,0 +1,9 @@
...
+  if (recipient) {
+    return isProperAddress(recipient) || isValidEnsName(recipient);
+  }
```
Comment (xitronix) [inline universal-login-commons/lib/core/utils/isValidRecipient.ts]: } else {

File: universal-login-react/src/ui/Transfer/Amount/TransferAmount.tsx [hunk 9]
```diff
@@ -1,75 +1,52 @@
...
+    <>
+      <TransferDropdown
+        sdk={deployedWallet.sdk}
+        tokenDetailsWithBalance={tokenDetailsWithBalance}
+        tokenDetails={tokenDetails}
+        setToken={(token: TokenDetails) => updateTransferDetailsWith({transferToken: token.address})}
+        className={transferAmountClassName}
+      />
+      <div className="transfer-amount-row">
+        <label className="transfer-amount-label" htmlFor="amount-eth">How much are you sending?</label>
```
Comment (xitronix) [inline universal-login-react/src/ui/Transfer/Amount/TransferAmount.tsx]: `<label>Amount</label>` will be constistent with 'Recipienist'

File: universal-login-react/src/ui/Transfer/Transfer.tsx [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+            tokenDetails={tokenDetails}
+            setAmountError={(isAmountInvalid: boolean) => setErrors({...errors, amountError: isAmountInvalid})}
+            amountError={errors.amountError}
+            transferAmountClassName={transferAmountClassName}
+          />
+          <TransferRecipient
+            updateTransferDetailsWith={updateTransferDetailsWith}
+            recipientError={errors.recipientError}
+            setRecipientError={(isRecipientInvalid: boolean) => setErrors({...errors, recipientError: isRecipientInvalid})}
+            transferAmountClassName={transferAmountClassName}
```
Comment (xitronix) [inline universal-login-react/src/ui/Transfer/Transfer.tsx]: Inconcistent name

File: universal-login-commons/lib/core/utils/isValidRecipient.ts (hunk 8 above)
Comment (cehali) [reply universal-login-commons/lib/core/utils/isValidRecipient.ts]: Why? It is not necessary here.

File: universal-login-react/src/ui/Transfer/Transfer.tsx (hunk 10 above)
Comment (cehali) [reply universal-login-react/src/ui/Transfer/Transfer.tsx]: 👍

File: universal-login-react/src/ui/Transfer/Transfer.tsx [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+            updateTransferDetailsWith={updateTransferDetailsWith}
+            tokenDetails={tokenDetails}
+            setAmountError={(isAmountInvalid: boolean) => setErrors({...errors, amountError: isAmountInvalid})}
+            amountError={errors.amountError}
+            transferAmountClassName={transferAmountClassName}
+          />
+          <TransferRecipient
+            updateTransferDetailsWith={updateTransferDetailsWith}
+            recipientError={errors.recipientError}
+            setRecipientError={(isRecipientInvalid: boolean) => setErrors({...errors, recipientError: isRecipientInvalid})}
```
[the remark this answers is not in the document]
Comment (cehali) [reply universal-login-react/src/ui/Transfer/Transfer.tsx]: They are slightly different: different margins, additional button or spans so I think it is much more effort than reward.

File: universal-login-react/src/ui/Transfer/Amount/TransferAmount.tsx (hunk 9 above)
Comment (cehali) [reply universal-login-react/src/ui/Transfer/Amount/TransferAmount.tsx]: 👍 Done

**Pull request merged**